src/main/java/net/neoforged/neoforge/network/registration/IPayloadRegistrar.java

```diff
@@ -0,0 +1,194 @@
+/*
+ * Copyright (c) NeoForged and contributors
+ * SPDX-License-Identifier: LGPL-2.1-only
+ */
+
+package net.neoforged.neoforge.network.registration;
+
+import java.util.function.Consumer;
+import net.minecraft.network.Connection;
+import net.minecraft.network.FriendlyByteBuf;
+import net.minecraft.network.protocol.common.custom.CustomPacketPayload;
+import net.minecraft.resources.ResourceLocation;
+import net.minecraft.server.network.ConfigurationTask;
+import net.neoforged.neoforge.network.handling.ConfigurationPayloadContext;
+import net.neoforged.neoforge.network.handling.IConfigurationPayloadHandler;
+import net.neoforged.neoforge.network.handling.IPayloadHandler;
+import net.neoforged.neoforge.network.handling.IPlayPayloadHandler;
+import net.neoforged.neoforge.network.handling.ISynchronizedWorkHandler;
+import net.neoforged.neoforge.network.handling.ITaskCompletedHandler;
+import net.neoforged.neoforge.network.handling.PlayPayloadContext;
+
+/**
+ * Defines a registrar for custom payloads that can be sent over the network.
+ * <p>
+ * A custom payload is a class which extends {@link CustomPacketPayload}, it is recommended to use a record for this.
+ * </p>
+ * <p>
+ * The payload is written to the networks outgoing buffer using the {@link CustomPacketPayload#write(FriendlyByteBuf)} method.
+ * However, to read the payload from the incoming buffer, your registered {@link FriendlyByteBuf.Reader} is used.
+ * <br>
+ * When you implement your {@link CustomPacketPayload#write(FriendlyByteBuf)} method you do not need to write the id of the payload,
+ * neither do you need to read it in your {@link FriendlyByteBuf.Reader} implementation. However, you do need to make sure that the
+ * id you pass into {@link #play(ResourceLocation, FriendlyByteBuf.Reader, IPlayPayloadHandler)} and
+ * {@link #configuration(ResourceLocation, FriendlyByteBuf.Reader, IConfigurationPayloadHandler)} is the same as the id you
+ * return from your {@link CustomPacketPayload#id()}. We suggest using a <code>public static final ResourceLocation</code> field
+ * to store it and then reference it in both places.
+ * <br>
+ * Ids can be reused between play and configuration payloads, but it is needed to use different ids for different payloads.
+ * <br>
+ * Under certain situations you are not able to register a payload:
+ * <ul>
+ * <li>If the id you are trying to register is already in use, meaning you used the same id twice for different packets of the same kind.</li>
+ * <li>If you are trying to register a payload to a namespace that is not your own.</li>
+ * </ul>
+ * In principal this means that the registration will fail if any of these cases occur.
+ * The exception thrown in these cases is a {@link RegistrationFailedException}.
```

Comment:
`principle` not `principal`. Also, this implies that there are cases where the registration will *not* fail? What are those and why would it not fail?